[#5] 유저 정보 업데이트 기능 구현

msugo1 opened the pull request:
## 📖 구현 항목
### 🔖 유저 정보 변경 기능관련 로직 구현
- 회원이 요청한 정보에 맞게 `현재 정보를 변경`하는 기능 구현 (변경 불가능 한 정보 제외 ex. ID, Password, Email 등)

### 🔖 유저 프로필 사진 추가/변경 기능관련 로직 구현
- 회원이 업로드 요청한 `프로필 사진을 추가/변경`하는 기능 구현
- 업로드 한 `파일`을 저장소에 `저장`하는 기능 구현.

 1. 파일이 한 곳에만 저장되는 것을 막기 위해 `유저의 ID를 반영`한 디렉토리 생성
 2. 파일 이름의 중복으로 기존 데이터가 변형되는 것을 막기 위해 `UUID`를 활용해 `랜덤한 숫자`형태로 `파일이름 변경`
 3. 파일 저장에 실패하면 제대로 된 업로드의 반영이 불가능하므로 예외 발생 - `FileNotUploadedException` 

- 불필요한 파일이 지속적으로 쌓이는 과정에서 `저장소의 용량 낭비를 막기 위해` 기존 프로필 사진 파일이 있다면 지우고 새로 요청 받은 파일을 저장하는 기능 구현

 1. 기존의 파일 삭제가 제대로 이루어지지 않은 경우 예외 발생 - `FileNotDeletedException` 

### 🔖 위의 로직을 테스트하기 위한 테스트 작성

Commit:
[#5] 유저 정보 업데이트 기능 구현(프로필 이미지 업로드 옵션 추가)

Model 관련
* user
- UserUpdateRequest (new)
: 사용자의 정보 변경 내용에 대한 요청을 담을 객체를 만들기 위한 클래스 생성

- UserUpdate (new)
: 업데이트 내용을 실제로 반영할 객체를 만들기 위한 클래스 생성
(기존 User 객체에서 변경 불가능한 부분 - ex. UserId, UserEmail - 을 제외한 내용을 담을 객체)

* file
- FileData (new)
: 프로필 사진에 대한 정보(좀더 일반화하면 파일에 대한 정보 - 이름과
경로)를 담기 위한 객체를 만드는 클래스 생성

Service 관련
* UserService (update)
- userInfoUpdate 메소드 추가
: 사용자 정보 변경요청을 처리하기 위한  비즈니스 로직 추가

* FileService (new)
- 파일 자체를 업로드/저장/삭제하는 것은 UserService와 다른 책임이라고
  생각하여 책임을 분리하기 위한 또 하나의 서비스 계층 생성
- 현재 수정 진행중으로 윤곽이 좀 더 잡히면 커밋에 업데이트 내역 반영할
  예정

Controller 관련
* Usercontroller (update)
- userInfoUpdate 메소드 임시 작성
: PostMan을 사용해 Controller가 들어오는 파일 업로드 요청을 제대로 처리하는지 테스트해보기 위해 임시적으로 추가한 메소드
(서비스 계층 구현 완료 후 수정 필요)

Mapper/Repository 관련
* UserMapper, UserRepository
- userUpdate 메소드 추가
: 서비스 계층에서 처리한 내용(유저 정보 업데이트)을 DB에 반영하기 위한
메소드

Commit:
[#5] 유저 정보 업데이트 기능 구현 (Update)

Model 관련
user

* UserUpdate (delete)
- 새로이 클래스 필요 없이 컨트롤러로부터 전달받은 UserUpdateRequest와 userId를 함께 DAO 계층으로 넘겨주면 될 듯 해서 Update 정보를 한곳에 모아주는 객체 생성 클래스를 삭제
(NOTE: controller 통합테스트 시 mapper 쪽에서 data integerity 관련 예외가 발생해서 일단 예외 해결 시도 해보고 안될 경우 다시 해당 클래스를 생성해서 trouble shooting 할 예정)

service 계층 관련

* FileService (update)
- path 생성, name 생성, 파일생성/제거, 파일 업로드에 관한 부분 각각 책임 분리 (테스트 작성 예정)

* UserService (update)
- 저장공간의 확보를 위해 사용자가 기존의 프로필 사진을 변경하는 경우
저장되어 있는 해당 파일을 삭제하고 나서 새롭게 파일 저장소에 업로드 된
프로파일 사진을 저장하는 로직 추가

- 파일업로드 실패에 관련한 IO예외를 런타임 예외로
   전환(FileUploadException)

controller 계층 관련

* UserController (update)
- 사용자의 정보를 업데이트 요청을 처리하는 메소드이름 변경
(userInfoUpdate -> userUpdate)

- 임시로 작성해 놓은 메소드 내용 수정

Repository/Mapper 관련

* UserRepository/UserMapper
- 사용자의 기존 프로파일 사진 삭제를 위해 사용자의 ID를 이용해 DB에
  저장된 파일의 이름과 경로를 가져오는 메소드를 추가

- UserService에서 사용자 정보를 업데이트하기 위해 DAO로 넘겨주는 파라미터의 변경으로 DAO 계층의  userUpdate 메소드 및 Query 수정

* 테스트 관련
UpdateTest
- 사용자 정보 변경을 위해 작성한 로직의 작동 여부를 테스트하기 위해 임시 테스트 작성

Commit:
[#5] 유저 정보 업데이트 기능 구현 (Update)

Service 계층 관련

* FileService (updte)
- IOException을 예외전환 (IOException -> FileNotUploaded/Deleted
  Exception)
: 기존에 IOException을 UserService로 던져서 UserService가 런타임 예외로
  전환하게 했던 것을 FileService에서 직접 전환하도록 함

* UserService (update)
- 기존 userUpdate 메소드의 책임 분리
: userUpdate가 프로필 이미지 업데이트와 유저 정보 업데이트를 모두
담당했던 것을 분리
(userUpdate -> userInfoUpdate (사용자 정보 업데이트 담당),
userProfileImageUpdate (사용자의 프로필 이미지 업데이트 담당)

controller 계층 관련

* UserController (update)
- 서비스 계층에서 유저 정보와 프로필 이미지를 업데이트하는 메소드들의
  책임 분리에 따라 Controller 계층에서도 요청 URL 분리

(userUpdate -> userProfileImageUpdate, userInfoUpdate)

Repository/Mapper 계층 관련

* UserRepository/UserMapper
- 위의 메소드 책임 분리에 따라 DAO 업데이트 반영 책임도 분리 (유저의
  정보 변경 반영 메소드 / 유저의 프로필 사진 변경 반영 메소드)
- 위의 메소드의 내용에 맞게 Query 수정

* 테스트 관련
- 추가/분리한 로직에 맞게 통합테스트 및 단위 테스트 작성 - 테스트 완료

f-lab-dev commented on src/main/java/me/soo/helloworld/controller/UserController.java:
```diff
@@ -72,9 +74,29 @@
-            userService.userPasswordUpdate(userId, userPasswordRequest);
-            loginService.logout();
+        userService.userPasswordUpdate(userId, userPasswordRequest);
+        loginService.logout();
```
`비밀번호 변경이 완료되면 로그아웃한다` 는 비지니스로직의 요구사항입니다. 그러면 여기에 있는게 맞을까요?

msugo1 replied:
비즈니스 로직이면 `service` 계층으로 가는게 맞겠네요! `UserController -> UserService`로 이동하였습니다.

f-lab-dev commented on src/main/java/me/soo/helloworld/controller/UserController.java:
```diff
@@ -72,9 +74,29 @@
+        return HTTP_RESPONSE_OK;
+    }
+
+    @PutMapping("/account/profile-image")
```
내 정보를 관리하는 Controller를 따로 분리하는게 낫지 않을까요?

msugo1 replied:
비밀번호, 회원정보, 프로필 사진 업데이트만 모아서 `UserUpdateController` 로 분리했습니다!

f-lab-dev commented on src/main/java/me/soo/helloworld/controller/UserController.java:
```diff
@@ -72,9 +74,29 @@
+            userService.userProfileImageUpdate(userId, profileImage);
             return HTTP_RESPONSE_OK;
+        } catch (FileNotUploadedException | FileNotDeletedException e) {
+            return HTTP_RESPONSE_INTERNAL_SERVER_ERROR;
```
에러코드는 리턴해주면서 로깅도 하는게 나아보입니다~
그리고 이런것은 공통적인 Exception을 만들어서 `ExceptionHandler`로 처리하는게 나아보이네요

msugo1 replied:
`ExceptionHandler`에 대해 좀 더 공부한 뒤 새로 이슈 하나 만들어서 업데이트 하도록 하겠습니다. 감사합니다.

msugo1 replied:
여러 자료를 보면서 배우느라 좀 오래 걸리긴 했지만  `ExceptionHandler` 사용해서 예외 처리하고 테스트까지 마무리했는데요. 문제는 제가 봤던 게시글(한글 영어 상관없이)마다 제각각에 표준이 정해져 있지 않아서 제대로 되었는지 모르겠네요. 혹시 다시 수정할 때 참고할 만한 문서나 게시글이 있을까요?

f-lab-dev replied:
표준은 원래 없긴 합니다. 결국 원리를 잘 알아내서 잘 조합하는 수밖에 없습니다~

f-lab-dev commented on src/main/java/me/soo/helloworld/service/FileService.java:
```diff
@@ -0,0 +1,91 @@
+import java.util.UUID;
+
+@Service
+public class FileService {
```
이 서비스는 `파일을 어디다 저장하냐`에 따라 확장의 여지가 있어보이네요

msugo1 replied:
`FileService`는 인터페이스로 만들고 저장하는 위치에 따라서 이름을 붙여 구현해서 확장성을 높이도로록 하겠습니다.

f-lab-dev commented on src/main/resources/application.properties:
```diff
@@ -2,4 +2,17 @@
+
+# 업로드 가능한 최대 파일 크기와 총 요청 크기설정 (단위 크기는 KB, MB 등을 숫자와 함께 지정가능하며 DataSize 클래스가 값을 변환하여 반환)
+# 조건을 초과한 요청이 들어오면 MultipartException 을 발생시킨다.
+spring.servlet.multipart.max-file-size=2MB
```
요새 카메라 화질이 좋아져서 꽤 용량이 큰데 이거로는 부족해보입니다~

msugo1 replied:
그러네요 제가 핸드폰에 찍어논 사진만 봐도 5~6MB는 기본으로 잡아먹네요;; 10MB로 바꿨는데 이 정도면 충분할지 모르겠네요

f-lab-dev commented on src/main/resources/mappers/UserMapper.xml:
```diff
@@ -3,24 +3,41 @@
+
+    <select id="getUserProfileImageById"
+            parameterType="String" resultType="me.soo.helloworld.model.file.FileData">
+        SELECT profileImageName, profileImagePath FROM users WHERE userId = #{userId}
```
쿼리에 적절하게 개행을 넣어주면 가독성을 높힐 수 있습니다

msugo1 replied:
추가하였습니다! 쿼리문은 가독성을 높일 생각을 못했었네요 감사합니다.

f-lab-dev commented on src/test/java/me/soo/helloworld/controller/UserControllerTest.java:
```diff
@@ -192,4 +204,62 @@ public void userPasswordUpdateTestSuccess() throws Exception {
+    @Test
+    @DisplayName("사용자의 프로파일 사진 업데이트에 성공하면 Http Status Code 200(OK)를 리턴합니다.")
+    public void userProfileUpdateTestSuccess() throws Exception {
+//        testUserSignUp(testUser);
```
이런 테스트용 코드는 push하지 않는게 좋습니다~

f-lab-dev reviewed (commented):
파일 업로드는 aws s3도 한번 고려해보시면 좋을 것 같네요~
중요한건 이런 솔루션을 왜 써야하는지, 무엇이 좋은지, 무엇이 안좋아지는지를 알아보시는거고요

Commit:
[#5] 피드백 반영

* UserMapper.xml
- 쿼리 가독성 향상을 위해 개행추가

* FileService 인터페이스 추가
* 기존 FileService 클래스는 위의 인터페이스를 구현하고
  LocalFileService로 변경
- 어디에 저장하느냐에 따라 확장가능성이 있기 때문에 Controller가
  인터페이스에 의존하도록 해서 OCD조건 충족

* UserUpdateController 추가
- 기존 UserController로부터 유저 비밀번호/정보/프로필 사진에 대한 요청을
  받는 부분 책임 분리

* ControllerExceptionHandler 추가
- Controller 내부의 메소드 userProfileImageUpdate에서 일일이 try/catch를
  통해 예외를 잡았던 부분을 ExceptionHandler를 사용해 외부로 분리시킴

- 성격이 비슷한 두예외- FileNotUploadedException,
  FileNotDeletedException - 가 FileException을 상속받게 해서 공통 예외를
  만들고 이를  ExceptionHandler 내의 매개변수를 이용해서
  ExceptionHandler가 두 예외를 한번에 처리할 수 있도록 함

* Multipart File 최대 Size 증가
- 좀더 고화질의 사진업로드 시 용량 고려

* 테스트 코드
- 불필요하게 담겨 있는 코드/주석이 없는지 검토 후 수정 혹은 삭제
- 책임이전에 따라 테스트 클래스도 분리
(UserControllerTest -> UserControllerTest / UserUpdateControllerTest

f-lab-dev commented on src/main/java/me/soo/helloworld/controller/UserUpdateController.java:
```diff
@@ -0,0 +1,51 @@
+
+    private final LoginService loginService;
+
+    @PutMapping("/account/update/password")
```
`PUT` 메소드 자체에 update라는 뜻이 있습니다~ 중복된 표현이 아닐까요?
이거는 이 문서를 자세히 보시면 좋을 것 같네요

https://docs.microsoft.com/ko-kr/azure/architecture/best-practices/api-design

msugo1 replied:
다시 한 번 꼼꼼하게 읽어봤는데요. 

혹시 

> 가능하다면 리소스 URI는 동사(리소스에 대한 작업)가 아닌 명사(리소스)를 기반으로 해야 합니다.

이 부분하고도 관련있는건가요? 일단 지적해주신데로 `PUT` 자체에 Update의 의미가 담겨있다는 것을 간과했네요. URL에서 update 제거했고 뒤에 명사만 남겨두었습니다.

f-lab-dev commented on src/main/java/me/soo/helloworld/controller/UserUpdateController.java:
```diff
@@ -0,0 +1,51 @@
+        userService.userPasswordUpdate(userId, userPasswordRequest);
+        loginService.logout();
+
+        return HTTP_RESPONSE_OK;
```
리턴타입을 void로 선언하면 이거는 생략이 가능합니다~

msugo1 replied:
Push 전에 불필요한 import, 설정들 제거한다고 수정한 클래스들 한번씩 쭉 훑어 봤는데 놓친 부분이 있었네요. `loginService.logout()` 부분은 Service 계층에서 처리하도록 바꿨기 때문에 컨트롤러에서 제거하였습니다.

f-lab-dev commented on src/main/java/me/soo/helloworld/util/ControllerExceptionHandler.java:
```diff
@@ -0,0 +1,25 @@
+    })
+    public ResponseEntity<String> fileException(final FileException ex) {
+        log.error("### FileException");
+        log.error("### FileException has occurred during uploading or deleting files");
```
이 내용은 에러를 디버깅하는데에 큰 도움이 되지 않습니다. `ex`를 같이 로깅하는게 좋아보이네요~

msugo1 replied:
error 메시지를 바꾸고 파라미터로 문자열 뿐만 아니라 ex 객체 또한 전달하는 것으로 수정했습니다!

f-lab-dev commented on src/main/resources/mappers/UserMapper.xml:
```diff
@@ -3,41 +3,60 @@
 
     <select id="getUserProfileImageById"
             parameterType="String" resultType="me.soo.helloworld.model.file.FileData">
+
```
취향차이지만 이런 곳에는 굳이 개행을 넣지 않아도 됩니다~

f-lab-dev commented on src/main/resources/mappers/UserMapper.xml:
```diff
@@ -3,41 +3,60 @@
     <select id="isUserIdDuplicate" parameterType="String" resultType="boolean">
+
         SELECT EXISTS
+
```
여기도 과도한 개행으로 오히려 가독성을 떨어뜨릴 수 있습니다~

msugo1 replied:
어렵네요; 쿼리문이 두 줄 이상 넘어가지 않는 경우는 개행이 불필요해보여서 다 제거하였고 좀 긴 쿼리문들만 개행 그대로 남겨두었습니다.

f-lab-dev replied:
이것은 스스로 계속 보면서 어디서 가독성이 떨어질 수 있을까를 계속 고민하셔야합니다~ 정해진 공식이 있는게 아니라요

Commit:
[#5] 피드백 반영

* UserMapper.xml
- Query문 가독성 향상을 위해 불필요한 개행 제거

* UserUpdateController
- PUT method와 중복되는 의미를 가진 URL 제거(/users/account/update/~ ->
  /users/account/~)

* UserController
- 불필요한 코드 제거
: return 값이 void인 loginService.logout() 가 컨트롤러에 남아있어서 제거

* ControllerExceptionHandler
- 에러 로깅 변경
: 로깅 메세지에 '예외 객체' 추가로 인자 전달

f-lab-dev commented on src/main/java/me/soo/helloworld/controller/UserUpdateController.java:
```diff
@@ -21,27 +21,24 @@
-    private final LoginService loginService;
-
-    @PutMapping("/account/update/password")
+    @PutMapping("/account/password")
```
이 클래스는 업데이트만 맡는 컨트롤러로 변경하셨는데, `내 정보를 관리하는 클래스`를 분리하기로 한 것 아니였을까요?

msugo1 replied:
이해를 잘못 하고 있던 부분을 바로잡고 내 정보를 관리하는 컨트롤러로 다시 분리완료하였습니다.

Commit:
[#5] 피드백 반영

* UserUpdateController -> MyPageController (변경)
- 단순히 모든 사용자의 정보를 업데이트하는 컨트롤러에서  현재 로그인 중인 사용자의 정보 관리를 담당하기 위한 컨트롤러로 변경

* UserService
- 현재 로그인 중인 사용자가 비밀번호를 변경했을 때 로그아웃시키는 로직
  제거

* 컨트롤러 명칭/책임 변경에 따른 테스트 이름 변경
- UserUpdateControllerTest -> MyPageControllerTest

f-lab-dev commented on src/main/java/me/soo/helloworld/controller/MyPageController.java:
```diff
@@ -16,32 +15,34 @@
 @RequiredArgsConstructor
-@RequestMapping("/users")
-public class UserUpdateController {
+@RequestMapping("/my-page")
```
이 곳은 복수형이여야 맞는 것 같네요~
그리고 내 정보라는 리소스를 표현하는 것이므로 `/my-infos`로 하시는게 나을 것 같아요

https://docs.microsoft.com/ko-kr/azure/architecture/best-practices/api-design

msugo1 replied:
꼼꼼히 읽는다고 읽어봤는데 아직도 리소스란 개념이 와닿지가 않네요 😢 피드백 반영했습니다. 감사합니다.

Commit:
[#5] 피드백 반영

* MyPageController
- URL이 가지는 '리소스'의 성격을 더 잘 표현하기 위해 URL 변경
(/my-page -> /my-infos)

- myInfoUpdate 메소드 URL 변경
(/infos -> /details)

f-lab-dev commented on src/main/java/me/soo/helloworld/controller/MyPageController.java:
```diff
@@ -38,7 +38,7 @@
     }
 
-    @PutMapping("/infos")
+    @PutMapping("/details")
```
내 정보를 전반적으로 업데이트하는거라면 `PUT /my-infos`로 충분하지 않을까요?

msugo1 replied:
넵 처음에 그렇게 생각했었는데 혹시나하는 마음에 뒤에 details를 넣었거든요. 제가 생각해도 굳이 필요 없는 부분을 덧붙인 것 같아서 제거하겠습니다.

Commit:
[#5] 피드백 반영

* MyPageController

myInfoUpdate
- 의미가 중복되어 불필요한 URL 제거
(/my-infos/details -> /my-infos)

msugo1 committed:
Merge branch 'develop' into feature/5
